Modules/_json.c

```diff
@@ -1643,19 +1689,30 @@ encoder_listencode_list(PyEncoderObject *s, _PyUnicodeWriter *writer,
 
     if (_PyUnicodeWriter_WriteChar(writer, '['))
         goto bail;
+
+    PyObject* separator = s->item_separator;
+    PyObject* newline_indent = 0;
     if (s->indent != Py_None) {
-        /* TODO: DOES NOT RUN */
-        indent_level += 1;
-        /*
-            newline_indent = '\n' + (' ' * (_indent * _current_indent_level))
-            separator = _item_separator + newline_indent
-            buf += newline_indent
-        */
+        indent_level++;
+        newline_indent = _create_newline_indent(s->indent, indent_level);
+        if (newline_indent == 0) {
+            goto bail;
+        }
+
+        if (_PyUnicodeWriter_WriteStr(writer, newline_indent)) {
+            goto bail;
+        }
+
+        separator = PyUnicode_Concat(separator, newline_indent);
```

Comment:
I think this leaks a reference to the previous contents of `separator`.

Reply:
The previous contents of separator was a borrowed reference, so it should be fine. But correct me if I am wrong!